docs_nnx/guides/flax_gspmd.md

```diff
@@ -0,0 +1,378 @@
+    )
+
+  def __call__(self, x: jax.Array):
+    y = self.dot1(x)
+    y = jax.nn.relu(y)
+    # In data parallelism, input / intermediate value's first dimension (batch)
+    # will be sharded on `data` axis
+    y = jax.lax.with_sharding_constraint(y, PartitionSpec('data', 'model'))
+    z = jnp.dot(y, self.w2.value)
+    return z
+```
+
+### Understand sharding names
+
+The so-called "sharding annotations" are essentially tuples of device axis names like `'data'`, `'model'` or `None`. This describes how each dimension of this JAX array should be sharded — either across one of the device mesh dimensions, or not sharded at all.
+
+So, when you define `W1` with shape `(depth, depth)` and annotate as `(None, 'model')`:
+
+* The first dimension will be replicated across all devices.
+* The second dimension will be sharded over the `'model'` axis of the device mesh. This means `W1` will be sharded 4-way on devices `(0, 4)`, `(1, 5)`, `(2, 6)` and `(3, 7)`, on this dimension.
+
+JAX's [Distributed arrays and automatic parallelization](https://jax.readthedocs.io/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html) guide offers more examples and explanations.
+
++++
+
+## Initialize a sharded model
+
+Now we have annotations attached to the `nnx.Variable`, but the actual weights haven't been sharded yet. If you just go ahead and create this model, all JAX arrays are still stuck in device 0. In practice, you'd want to avoid this, because a large model will OOM in this situation, and all the other devices are not utilized.
+
+```{code-cell} ipython3
+unsharded_model = DotReluDot(1024, rngs=nnx.Rngs(0))
+
+# We have annotations sticked there, yay!
+print(unsharded_model.dot1.kernel.sharding)     # (None, 'model')
+print(unsharded_model.w2.sharding)              # ('model', None)
+
+# But the actual arrays are not sharded... wut?
+print(unsharded_model.dot1.kernel.value.sharding)  # SingleDeviceSharding
+print(unsharded_model.w2.value.sharding)           # SingleDeviceSharding
+```
+
+We should leverage JAX's compilation mechanism, aka. `jax.jit`, to create the sharded model. The key is to intialize a model and assign shardings upon the model state within a jitted function:
+
+1. Use [`nnx.get_partition_spec`](https://flax-nnx.readthedocs.io/en/latest/api_reference/flax.nnx/spmd.html#flax.nnx.get_partition_spec) to strip out the `.sharding` annotations attached upon model variables;
+
+1. Call [`jax.lax.with_sharding_constraint`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.with_sharding_constraint.html) to bind the model state with the sharding annotations. This API tells the top-level `jax.jit` how to shard a variable!
+
+1. Throw away the unsharded state and return the model based upon the sharded state.
+
+1. Compile the whole function with `nnx.jit` instead of `jax.jit` because it allows the output to be a stateful NNX module.
+
+1. Run it under a device mesh context so that JAX knows which devices to shard it to.
```

Comment:
```suggestion
1. Call [`jax.lax.with_sharding_constraint`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.with_sharding_constraint.html) to bind the model state with the sharding annotations. This API tells the top-level `nnx.jit` how to shard a variable!

1. Throw away the unsharded state and return the model based upon the sharded state.

1. Run it under a device mesh context so that JAX knows which devices to shard it to.
```

Reply:
Hmm... I think we should still briefly explain why using `nnx.jit` is a better pattern. Especially since we are making transforms closer to JAX style now we should assume some users have experience with `jax.jit`. I can remove the mentions of `jax.jit` here and direct users more explicitly to `nnx.jit`.